### ISSUE TITLE
Enable Customization of IMPORTED_NAMES

### ISSUE THREAD

**Issue by baer** [formatjs/babel-plugin-react-intl#17]
First, after a few tries at i18n in various systems I have to say that this auto extraction plugin is an AWESOME IDEA. I'm very excited to have a translation toolchain rather than just a bunch of message properties files.

I'd like to request (and will happily PR) a small new feature: User defined extensions to [IMPORTED_NAMES](https://github.com/yahoo/babel-plugin-react-intl/blob/master/src/index.js#L22). 

---

Basically the use case is this: I have an app with both locale specific translations and tenant specific messages. For example in one tenant a message might read: "Hello and welcome to Brand1" and the other might be "Hello and welcome to Brand2". This means that translated files will be the number of supported tenants \* number of supported locales which makes this a 2 dimensional lookup (brand and locale) rather than a 1 dimensional one (locale).

brand1-en-US.json
brand1-en-CA.json
brand2-en-US.json
brand2-en-CA.json

I have been doing a lot of thinking about how to effectively manage/extract this information without a huge burden on the developers and here is what I've come up with:

``` js
// Define co-located brand overrides by prefixing the ID with the brand name.
// Sorting/filtering will be a post processing step
var tenantOverrides = defineMessages({
  id: "brand2.my-message"
  default: "Hooray I'm in Brand 2!"
})

// Any use of formatMessage etc will be for the base brand
var message = formatMessage({
  id: "my-message"
  default: "Hooray I'm in Brand 1!"
})
```

If I had this feature, I would add `defineBrandOverrides` to use in place of `defineMessages` when that is what I am doing. This allows me to let developers continue to co-locate all untranslated messages and keep using the message extractor as is. It also means that if there are no overrides there is no additional work.

**Comment by ericf** [comment]
The imported names relate specifically to things React Intl `export`s. The plugin then uses these to check whether an identifier references an import of one of React Intl's exports.

Can you treat the "brand" as data? Presumably the brand name wouldn't be translated, but if it was, the following could still work:

```js
import {
    defineMessages,
    FormattedMessage,
} from 'react-intl';

const strings = defineMessages({
    greeting: `Hooray I'm in {brand}`
});

export default const Component = ({brand}) => (
    <FormattedMessage
        {...strings.greeting}
        values={{brand}}
    />
);
```

Is there a reason you can't use a [simple argument](http://formatjs.io/guides/message-syntax/#simple-argument) in the ICU Message?

**Comment by baer** [comment]
@ericf Thanks for the response! The problem with using a contrived example is not being able to explain the problem fully. Our apps (walmartlabs) are deployed in ~25 different markets. Each of these markets then further localizes their apps. It is frequently the case that one market (what I called a brand above) has completely different text as another. For example the reviews section of a product in one brand might be "Customer reviews" and in the other it might be "Our Customers Say".

So now we have the following problem: How do you write code that can be run in multiple markets (and localizations), doesn't introduce huge developer burden AND reduces branching/if statements based on brand as much as possible. 

Right now, we use a directory structure that is VERY similar to the react-native world where only UI that needs to change per platform gets a special directory and as much code as possible gets pushed into shared code.

```
common-widget/
- brand1-customizations
- brand2-customizations
- common
```

Formatted messages in the custom UI sections are easy but what about code in common? Form validation for example may have common JS/UI but need different message strings per brand. We have 3 bad options:

1. We ask developers to maintain and keep in sync by hand all of the override message strings that this babel plugin generates for the default - This works and it maintains the purity of common code BUT as you pointed out in the RFC for react-intl@2.x keeping this in sync will suck and be error prone. It also forces all developers to understand i18n a bit so they don't mess it up accidentally.
2. Co-locate the strings ( :confetti_ball: ) using defineMessages and asking devs to prefix their override keys. If we do this we can generate the proper `messages.json` for each brand at build time at the cost of some purity (but no branching) in common code.
3. Take step2 and allow the use of custom defineMessages statements such as defineBrandOverrides. This would allow us to reduce the cognitive load for new developers. If we took this even further and allow this plugin to auto-prefix messages we'd eliminate the need to ask devs to prefix keys themselves. That part is a nice-to-have :).

Basically, this is a small feature that could dramatically improve developer ergonomics for projects of this size.

As I mentioned before - I am happy to submit a PR for this! You've already gotten it so far :smile:

**Comment by ericf** [comment]
> Formatted messages in the custom UI sections are easy but what about code in common?

Saying something is "common" is a signal to me that it lacks enough context to be able to know and define the custom brand-related overrides. Instead of having these common components become "fat" with all possible overrides, wouldn't it be better to have them _take_ in a prop, like a React Intl Message Descriptor?

```js
import React, {PropTypes} from 'react';
import {
    defineMessages,
    FormattedMessage,
} from 'react-intl';

const strings = defineMessages({
    greeting: {
        id: 'greeting',
        defaultMessage: 'Hello, {name}!',
    },
});

const CommonComponent = ({name, greetingMessage}) => (
    <FormattedMessage
        {...(greetingMessage || strings.greeting)}
        values={{name}}
    />
);

CommonComponent.propTypes = {
    name: PropTypes.string.isRequired,
    greetingMessage: PropTypes.shape({
        id: PropTypes.string.isRequired,
        defaultMessage: PropTypes.string,
    }),
};

export default CommonComponent;
```

Are most things overridden, or only a few? If it's just a few key strings, like your example of "Customer Reviews", then is having the common component be customizable via its `props` reasonable?

The trouble with your idea of configuring which imported names to extract messages from is that you actually have to `import {someCustomThingThatDoesNotExist} from 'react-intl'`.

**Comment by baer** [comment]
I understand where you're coming from w/r to the "common" things lacking context and for the most part I think you're right with the exception of two caveats.

1. **common doesn't mean generic** - We have a UX team that defines the "global" way of doing things. In fact one of the best things about our migration to react is the ability to wrap up this "blessed" way of doing things and let teams integrate it easily as a component. Anyhow, these components often DO have enough context to define their own messages.
2. **In the absence of language enforced guarantee, mistakes will happen** - There are well over 100 developers at all different skill levels consuming these components. As much as I want to make sure every line is DRY, well tested, etc., developers under pressure without a simple and prescriptive path forward will often take shortcuts - the result of which I have to fix (sometimes late at night and using an IE8 VM :frowning: ). For that reason I'm trying to keep things as simple as possible which in this case may not be quite as clean from a CS standpoint.

I am about to start work on that will give me more context into this specific problem. If you can give me a week or two I can re-post here and give a more concrete example or suggestion about how this might work.

**Comment by baer** [comment]
> The trouble with your idea of configuring which imported names to extract messages from is that you actually have to import {someCustomThingThatDoesNotExist} from 'react-intl'

:point_up: that's true - I am going to have to think about that.

**Comment by longlho** [comment]
close due to stale
